### DISCUSSION TITLE
Body-level factorization for computing Coriolis matrix (also for the case of Christoffel consistent version)

### DISCUSSION THREAD

**Discussion by sentojh** [stack-of-tasks/pinocchio#2427] [General]
Hi, there.

I'm wondering about the algorithm to compute the Coriolis matrix, especially in [Alg. 1, [Numerical Methods to Compute the Coriolis Matrix and Christoffel Symbols for Rigid-Body Systems](https://asmedigitalcollection.asme.org/computationalnonlinear/article/16/9/091004/1109302/Numerical-Methods-to-Compute-the-Coriolis-Matrix)] and [Alg. 1, [Coriolis Factorizations and their Connections to Riemannian Geometry](https://arxiv.org/abs/2312.14425)], which present body-level factorizations to satisfy skew-symm. and also describe the Christoffel consistent factorization @pwensing.

I think both described algorithms use the same factorization, which is $B_i = \frac{1}{2} ((v_i \times^* )I_i + (I_i v_i \overline{\times}^* ) - I_i (v_i \times)) $. **Is it right to understand the above factorization is actually Christoffel consistent factorizaiton?** (The above factorization is the past version in Pinocchio library, and I checked the extension following the Chriostoffel consistent version, but I cannot find any difference.).

I'm a little confused about distinguishing between both papers in view of the implemetation of **algorithm to obtain coriolis matrix with specific body-level factorization**. Especially, is the matlab code [(link)](https://github.com/ROAM-Lab-ND/spatial_v2_extended/blob/main/v3/dynamics/CoriolisMatrix.m) from github - @pwensing exactly the same in [Alg. 1, [Numerical Methods to Compute the Coriolis Matrix and Christoffel Symbols for Rigid-Body Systems](https://asmedigitalcollection.asme.org/computationalnonlinear/article/16/9/091004/1109302/Numerical-Methods-to-Compute-the-Coriolis-Matrix)]?? I also read the discussion about in #1663, but still confused for me.

Thank you for your attention.

**Comment by pwensing** [comment]
Hi Jonghyeok,

The implementation of the updated factorization in Pinocchio is summarized in this PR: #1665 -- I believe that prior to this PR, Pinocchio had adopted the body-level factorization $B = v \times^* I$. Justin had found the utility of this factorization during his first-order derivatives work -- it's perfectly valid for getting a Coriolis matrix that satisfies $\dot{H}= C+C^T$ as is needed for passivity-based control -- the updated one just opens up a few more use cases, as discussed in #1663.

Regarding the algorithm in the papers: It's the same algorithm, but the second paper certifies its properties for a wider range of systems (the Christoffel-consistent property was originally only certified for mechanisms with revolute/prismatic joints, while the second paper shows this property when including floating-base points, joints with configuration-dependent joints axes [e.g., universal joints], or for closed-chain systems represented in tree-structure form using Jain's constraint embedding ideas). These details aren't really Pinocchio-specific, so I'm happy to move any further clarifying questions to email.

I hope this helps bring clarity.

Best,
-Pat

**Reply by sentojh** [reply]
This question might be rude, but thanks for your kindness and support!

I understand your explanation.

The only thing that I still need clarification on is the notation for describing rigid body dynamics using Spatial Vector Algebra (SVA) since I need to familiarize myself with this notation, which is closely related to Featherstone's notation. I'll try to understand more and more.

Anyway, if I have any further clarifying questions about the paper, I'll email them to you.

Best regards,

Jonghyeok
